design-proposals/migration-target.md

```diff
@@ -0,0 +1,97 @@
+
+This means that, only for VMs with LiveUpdate rollout strategy, we can already force the target for a live migration with something like:
+- set a (temporary?) nodeSelector/affinity on the VM
+- wait for it to be propagated to the VMI due to LiveUpdate rollout strategy
+- trigger a live migration with existing APIs (no need for any code change)
+- wait for the migration to complete
+- (eventually) remove the (temporary?) nodeSelector to let the VM be freely migrate to any node in the future
+
+Such a flow can already be implemented today with a pipeline or directly from a client like `virtctl` without any backend change.
+The drawback of that strategy is that we should tolerate having the spec of the VM amended twice with an unclear boundary about what was asked by the VM owner for long-lasting application specific reasons and what is required by a maintenance operator just for a specific migration attempt.
+
+## API Examples
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstanceMigration
+metadata:
+  name: migration-job
+spec:
+  vmiName: vmi-fedora
+  nodeName: my-new-target-node
+```
+
+## Scalability
+Forcing a `nodeName` on `VirtualMachineInstanceMigration` will cause it to be propagated to the destination virt-launcher pod. Having a `nodeName` on a pod will bypass the k8s scheduler and this could potentially lead to an unbalanced node placement across the nodes.
+But the same result can be already achieved today specifying a `nodeSelector` or `affinity` and `anti-affinity` rules on a VM. 
+
+## Update/Rollback Compatibility
+`nodeName` on `VirtualMachineInstanceMigration` will be only an optional field so no impact in terms of update compatibility.
+
+## Functional Testing Approach
+- positive test 1: a VirtualMachineInstanceMigration with an explict nodeName pointing to a node able to accommodate the VM should succeed
+- positive test 2: a VirtualMachineInstanceMigration with an explict nodeName pointing to a node able to accommodate the VM but not matching a nodeSelector already present on the VM should succeed
+- negative test 1: a VirtualMachineInstanceMigration with an explict nodeName should be refused if the required node doesn't exist
+- negative test 2: a VirtualMachineInstanceMigration with an explict nodeName should be refused if the VM is already running on the requested node
+- negative test 3: a VirtualMachineInstanceMigration with an explict nodeName should be refused if the user is not allowed to list nodes in the cluster
+- negative test 4: a VirtualMachineInstanceMigration with an explict nodeName should fail if the selected target node is not able to accommodate the additional pod for virt-launcher
+
+# Implementation Phases
+A really close attempt was already tried in the past with https://github.com/kubevirt/kubevirt/pull/10712 but the Pr got some pushbacks.
+A similar PR should be reopened, refined and we should implement functional tests.
```

Comment:
Would you outline the nature of the pushback? Do we currently have good answers to the issues raised back then?

Reply:
done, thanks

Reply:
Trying to summarize (@EdDev please keep me honest on this), it was somehow considered a semi-imperative approach and it was pointed out that a similar behavior could already indirectly be achieved modifying on the fly and then reverting affinity rules on the VM object.
see: https://github.com/kubevirt/kubevirt/pull/10712#issuecomment-1813963023
and: https://github.com/kubevirt/kubevirt/pull/10712#issuecomment-2144957961

How much this is imperative is questionable: at the end we already have a `VirtualMachineInstanceMigration` object that you can use to declare that you want to trigger a live migration, this is only about letting you also declare that you want to have a live migration to a named host.

The alternative approach based on amending the affinity rules on the VM object and waiting for the LiveUpdate rollout strategy to propagate it to the VMI before trying a live migration is described, pointing out its main drawback, in the `Alternative design` section in this proposal.

Reply:
Can you inline this succinctly? E.g, `that Pr got some pushbacks because it was not clear why a new API for one-off migration is needed. We give here a better explanation why this one-off migration destination request is necessary`.

Reply:
done

Reply:
- The "one-time" operation convinced me.
- The reasoning for the real need is hard for me, but I did feedback on this proposal what is convincing me.